File: scripts/beagle_config.py
#!/usr/bin/env python3
"""
BEAGLE Configuration Module for Python
Centralized configuration management respecting SAFE_MODE and BEAGLE_DATA_DIR
"""

import os
from pathlib import Path
from typing import Optional


def _bool_env(name: str, default: bool) -> bool:
    """Helper para ler variáveis de ambiente booleanas."""
    val = os.getenv(name)
    if val is None:
        return default
    val = val.strip().lower()
    return val in ("1", "true", "t", "yes", "y")


def safe_mode() -> bool:
    """
    Verifica se SAFE_MODE está ativo (default: True).
    
    Quando ativo:
    - Nenhuma publicação real será feita (arXiv, Twitter, etc.)
    - Ações críticas são logadas mas não executadas
    """
    return _bool_env("BEAGLE_SAFE_MODE", True)


def beagle_data_dir() -> Path:
    """
    Retorna o diretório base de dados do BEAGLE.
    
    Ordem de prioridade:
    1. Variável de ambiente BEAGLE_DATA_DIR
    2. ~/beagle-data (padrão)
    """
    if "BEAGLE_DATA_DIR" in os.environ:
        return Path(os.environ["BEAGLE_DATA_DIR"])
    return Path.home() / "beagle-data"


def data_dir() -> Path:
    """Alias para beagle_data_dir()."""
    return beagle_data_dir()


def models_dir() -> Path:
    """Path para modelos LLM."""
    return beagle_data_dir() / "models"


def logs_dir() -> Path:
    """Path para logs."""
    return beagle_data_dir() / "logs"


def papers_drafts_dir() -> Path:
    """Path para drafts de papers."""
    return beagle_data_dir() / "papers" / "drafts"


def papers_final_dir() -> Path:
    """Path para papers finais."""
    return beagle_data_dir() / "papers" / "final"


def vllm_url() -> str:
    """URL do servidor vLLM (default: http://t560.local:8000)."""
    return os.getenv("BEAGLE_VLLM_URL", "http://t560.local:8000")


def arxiv_token() -> Optional[str]:
    """Token da API arXiv (opcional)."""
    return os.getenv("ARXIV_API_TOKEN")


def publish_mode() -> str:
    """
    Retorna o modo de publicação atual.
    
    Valores:
    - "dry" (default) - nunca chama API real
    - "manual" - exige confirmação humana
    - "auto" - permite publicação automática (só se safe_mode() == False)
    """
    mode = os.getenv("BEAGLE_PUBLISH_MODE", "dry").lower().strip()
    if mode not in ("dry", "manual", "auto"):
        return "dry"
    return mode


def can_publish_real() -> bool:
    """
    Verifica se publicação real é permitida.
    
    Retorna True apenas se:
    - safe_mode() == False
    - E publish_mode() == "auto"
    """
    return not safe_mode() and publish_mode() == "auto"


# Exemplo de uso em funções de publicação
def publish_paper(paper_title: str, paper_content: str, dry_run_override: bool = False) -> bool:
    """
    Exemplo de função de publicação que respeita SAFE_MODE.
    
    Args:
        paper_title: Título do paper
        paper_content: Conteúdo do paper
        dry_run_override: Força dry-run mesmo se config permitir publicação
    
    Returns:
        True se publicado (ou simulado), False se bloqueado
    """
    if safe_mode() or dry_run_override:
        print(f"SAFE_MODE/dry-run: não vou enviar paper real, apenas salvar plano.")
        plan_path = papers_drafts_dir() / f"{paper_title.replace(' ', '_')}_plan.json"
        plan_path.parent.mkdir(parents=True, exist_ok=True)
        with open(plan_path, "w") as f:
            f.write(f'{{"title": "{paper_title}", "mode": "dry-run"}}')
        print(f"Plano salvo em: {plan_path}")
        return False
    
    if not can_publish_real():
        print(f"Publicação bloqueada: safe_mode={safe_mode()}, publish_mode={publish_mode()}")
        return False
    
    # Aqui iria a chamada real da API arXiv
    print(f"Publicando paper: {paper_title}")
    return True


if __name__ == "__main__":
    # Testes básicos
    print(f"SAFE_MODE: {safe_mode()}")
    print(f"Data dir: {beagle_data_dir()}")
    print(f"Models dir: {models_dir()}")
    print(f"Publish mode: {publish_mode()}")
    print(f"Can publish real: {can_publish_real()}")

File: beagle-julia/src/BeagleConfig.jl
# BEAGLE Configuration Module for Julia
# Centralized configuration management respecting SAFE_MODE and BEAGLE_DATA_DIR

module BeagleConfig

export beagle_data_dir, safe_mode, models_dir, logs_dir, papers_drafts_dir, papers_final_dir

"""
    beagle_data_dir() -> String

Retorna o diretório base de dados do BEAGLE.

Ordem de prioridade:
1. Variável de ambiente BEAGLE_DATA_DIR
2. ~/beagle-data (padrão)
"""
function beagle_data_dir()
    if haskey(ENV, "BEAGLE_DATA_DIR")
        return ENV["BEAGLE_DATA_DIR"]
    end
    return joinpath(homedir(), "beagle-data")
end

"""
    safe_mode() -> Bool

Verifica se SAFE_MODE está ativo (default: true).

Quando ativo:
- Nenhuma publicação real será feita (arXiv, Twitter, etc.)
- Ações críticas são logadas mas não executadas
"""
function safe_mode()
    val = get(ENV, "BEAGLE_SAFE_MODE", "true")
    val_lower = lowercase(strip(val))
    return val_lower in ("1", "true", "t", "yes", "y")
end

"""
    models_dir() -> String

Path para modelos LLM.
"""
function models_dir()
    return joinpath(beagle_data_dir(), "models")
end

"""
    logs_dir() -> String

Path para logs.
"""
function logs_dir()
    return joinpath(beagle_data_dir(), "logs")
end

"""
    papers_drafts_dir() -> String

Path para drafts de papers.
"""
function papers_drafts_dir()
    return joinpath(beagle_data_dir(), "papers", "drafts")
end

"""
    papers_final_dir() -> String

Path para papers finais.
"""
function papers_final_dir()
    return joinpath(beagle_data_dir(), "papers", "final")
end

"""
    publish_mode() -> Symbol

Retorna o modo de publicação atual.

Valores:
- :dry (default) - nunca chama API real
- :manual - exige confirmação humana
- :auto - permite publicação automática (só se safe_mode() == false)
"""
function publish_mode()
    mode_str = lowercase(strip(get(ENV, "BEAGLE_PUBLISH_MODE", "dry")))
    if mode_str == "auto"
        return :auto
    elseif mode_str == "manual"
        return :manual
    else
        return :dry
    end
end

"""
    can_publish_real() -> Bool

Verifica se publicação real é permitida.

Retorna true apenas se:
- safe_mode() == false
- E publish_mode() == :auto
"""
function can_publish_real()
    return !safe_mode() && publish_mode() == :auto
end

end # module

